plugin/restore/mysql/mysql.go

```diff
@@ -520,9 +520,45 @@ func binlogFilesAreContinuous(files []BinlogFile) bool {
 	return true
 }
 
-// FetchArchivedBinlogFiles downloads the missing binlog files from the remote instance to `binlogDir`,
-// but exclude latest binlog. We may  download the latest binlog only when doing PITR.
-func (r *Restore) FetchArchivedBinlogFiles(ctx context.Context) error {
+// Download binlog files on server.
+func (r *Restore) downloadBinlogFilesOnServer(ctx context.Context, binlogFilesLocal, binlogFilesOnServerSorted []BinlogFile) error {
+	if len(binlogFilesOnServerSorted) == 0 {
+		return fmt.Errorf("no binlog files on server")
+	}
+	latestBinlogFileOnServer := binlogFilesOnServerSorted[len(binlogFilesOnServerSorted)-1]
+	binlogFilesLocalMap := make(map[string]BinlogFile)
+	for _, file := range binlogFilesLocal {
+		binlogFilesLocalMap[file.Name] = file
+	}
+	log.Debug("Downloading binlog files", zap.Array("fileList", ZapBinlogFiles(binlogFilesOnServerSorted)))
+	for _, fileOnServer := range binlogFilesOnServerSorted {
+		fileLocal, existLocal := binlogFilesLocalMap[fileOnServer.Name]
+		path := filepath.Join(r.binlogDir, fileOnServer.Name)
+		if !existLocal {
+			if err := r.downloadBinlogFile(ctx, fileOnServer, fileOnServer.Name == latestBinlogFileOnServer.Name); err != nil {
+				log.Error("Failed to download binlog file", zap.String("path", path), zap.Error(err))
+				return fmt.Errorf("failed to download binlog file[%s], error[%w]", path, err)
+			}
+		} else if fileLocal.Size != fileOnServer.Size {
+			log.Debug("Deleting inconsistent local binlog file",
+				zap.String("path", path),
+				zap.Int64("sizeLocal", fileOnServer.Size),
+				zap.Int64("sizeOnServer", fileOnServer.Size))
+			if err := os.Remove(path); err != nil {
+				log.Error("Failed to remove inconsistent local binlog file", zap.String("path", path), zap.Error(err))
+				return fmt.Errorf("failed to remove inconsistent local binlog file[%s], error[%w]", path, err)
+			}
+			if err := r.downloadBinlogFile(ctx, fileOnServer, fileOnServer.Name == latestBinlogFileOnServer.Name); err != nil {
```

Comment:
This can be shared.

```
download := !existLocal
if existLocal && fileLocal.Size != fileOnServer.Size {
  // delete
  download = true
}
if download {
  ...
}

```

Reply:
The original intention was to provide different error messages for different downloading reasons.